dive_core/command_hierarchy.cpp

```diff
@@ -2089,32 +2093,157 @@ void CommandHierarchyCreator::AppendPacketFieldNodes(const IMemoryManager &mem_m
+                AddChild(CommandHierarchy::kSubmitTopology, packet_node_index, field_node_index);
+                AddChild(CommandHierarchy::kAllEventTopology, packet_node_index, field_node_index);
+                AddChild(CommandHierarchy::kRgpTopology, packet_node_index, field_node_index);
+            }
+        }
+    }
+}
 
-            CommandHierarchy::AuxInfo aux_info = CommandHierarchy::AuxInfo::RegFieldNode(
-            is_ce_packet);
-            uint64_t field_node_index = AddNode(NodeType::kFieldNode,
-                                                field_string_stream.str(),
-                                                aux_info);
+//--------------------------------------------------------------------------------------------------
+void CommandHierarchyCreator::AppendLoadStateExtBufferNode(const IMemoryManager &mem_manager,
+                                                           uint32_t              submit_index,
+                                                           uint64_t              va_addr,
+                                                           uint64_t              packet_node_index)
+{
+    PM4_CP_LOAD_STATE6 packet;
+    DIVE_VERIFY(mem_manager.CopyMemory(&packet, submit_index, va_addr, sizeof(packet)));
 
-            // Add it as child to packet_node
-            AddChild(CommandHierarchy::kEngineTopology, packet_node_index, field_node_index);
-            AddChild(CommandHierarchy::kSubmitTopology, packet_node_index, field_node_index);
-            AddChild(CommandHierarchy::kAllEventTopology, packet_node_index, field_node_index);
-            AddChild(CommandHierarchy::kRgpTopology, packet_node_index, field_node_index);
+    enum class StateBlockCat
+    {
+        kTex,
+        kShader,
+        kIbo
+    };
+    StateBlockCat cat;
+    switch (packet.bitfields0.STATE_BLOCK)
+    {
+    case SB6_VS_TEX:
+    case SB6_HS_TEX:
+    case SB6_DS_TEX:
+    case SB6_GS_TEX:
+    case SB6_FS_TEX:
+    case SB6_CS_TEX: cat = StateBlockCat::kTex; break;
+    case SB6_VS_SHADER:
+    case SB6_HS_SHADER:
+    case SB6_DS_SHADER:
+    case SB6_GS_SHADER:
+    case SB6_FS_SHADER:
+    case SB6_CS_SHADER: cat = StateBlockCat::kShader; break;
+    case SB6_IBO:
+    case SB6_CS_IBO:
+        DIVE_ASSERT("These seem to be specific to pre-A6XX");
+        cat = StateBlockCat::kIbo;
+        break;
+    default: DIVE_ASSERT(false); cat = StateBlockCat::kTex;
+    }
+
+    uint64_t ext_src_addr = 0;
+    switch (packet.bitfields0.STATE_SRC)
+    {
+    case SS6_DIRECT: ext_src_addr = va_addr + sizeof(PM4_CP_LOAD_STATE6); break;
+    case SS6_BINDLESS: DIVE_ASSERT("Not supported yet"); break;
+    case SS6_INDIRECT:
+        ext_src_addr = packet.u32All1 & 0xfffffffc;
+        ext_src_addr |= ((uint64_t)packet.u32All2) << 32;
+        break;
+    }
+
+#define APPEND_SHARPS(Type)                                                  \
```

Comment:
is there a specific reason why you prefer macro here? sometimes i found it is a bit more difficult to debug macros. maybe better with lambda expression for a short func like this?

Reply:
Lambda tends to be difficult to debug and slow due to Windows-STL issues. So I tend to avoid lambdas for short code snippets. But you're right, there's enough code here that it might make sense to so we can step into the code if needed.

Reply:
Done!

Reply:
sorry maybe i am not seeing the right revision, but it seems that the macro is still there

Reply:
I must've messed up. Will upload again tonight.